src/main/java/hudson/plugins/sonar/client/WsClient.java

```diff
@@ -48,17 +51,43 @@ public WsClient(HttpClient client, String serverUrl, String username, String pas
     this.password = password;
   }
 
-  public CETask getCETask(String taskId) throws Exception {
+  public static WsClient create(HttpClient client, SonarInstallation inst) {
+    String serverUrl = StringUtils.isEmpty(inst.getServerUrl()) ? SonarInstallation.DEFAULT_SERVER_URL : inst.getServerUrl();
+    if (StringUtils.isNotEmpty(inst.getServerAuthenticationToken())) {
+      return new WsClient(client, serverUrl, inst.getServerAuthenticationToken(), null);
+    } else {
+      return new WsClient(client, serverUrl, inst.getSonarLogin(), inst.getSonarPassword());
+    }
+  }
+
+  public CETask getCETask(String taskId) {
     String url = serverUrl + API_CE_TASK + taskId;
     String text = client.getHttp(url, username, password);
+    try {
+      JSONObject json = (JSONObject) JSONSerializer.toJSON(text);
+      JSONObject task = json.getJSONObject("task");
+
+      String status = task.getString("status");
+      String componentName = task.getString("componentName");
+      String componentKey = task.getString("componentKey");
+      // No analysisId if task is pending
+      String analysisId = task.optString("analysisId", null);
+      return new CETask(status, componentName, componentKey, url, analysisId);
+    } catch (JSONException e) {
+      throw new IllegalStateException("Unable to parse response from " + url + ":\n" + text);
+    }
+
+  }
+
+  public ProjectQualityGate getQualityGateWithAnalysisId(String analysisId) {
+    String url = serverUrl + API_PROJECT_STATUS_WITH_ANALYSISID + encode(analysisId);
+    String text = client.getHttp(url, username, password);
     JSONObject json = (JSONObject) JSONSerializer.toJSON(text);
-    JSONObject task = json.getJSONObject("task");
+    JSONObject projectStatus = json.getJSONObject("projectStatus");
 
-    String status = task.getString("status");
-    String componentName = task.getString("componentName");
-    String componentKey = task.getString("componentKey");
+    String status = projectStatus.getString("status");
 
-    return new CETask(status, componentName, componentKey, url);
+    return new ProjectQualityGate(null, status);
   }
 
   public ProjectQualityGate getQualityGate54(String projectKey) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00112)